WP_SW_Manager.php

```diff
@@ -2,195 +2,195 @@
+		 * with this function.
+		 *
+		 * @api
+		 * @param string $scope The scope identifying the service worker.
+		 * If omitted, it defaults to the URL where WordPress site is installed.
+		 * @return string A string representing the unique identifier for the
+		 * service worker registration promise.
+		 */
+		public function sw_js_id( $scope = '' ) {
+			if ( ! $scope ) { $scope = $this->default_scope(); }
+			return $scope;
+		}
+
+		private function default_scope() {
+			return site_url( '/', 'relative' );
+		}
+
+		private function setup_sw_registrar_script() {
+			$this->dynamic_server->add_file( self::SW_REGISTRAR_SCRIPT_URL,  array( $this, 'sw_registrar' ) );
+			add_action( 'init', array( $this, 'check_registrations' ), 999 );
+			add_action( 'wp_enqueue_scripts', array( $this, 'enqueue_registrar' ) );
+		}
+
+		public function check_registrations() {
+			$last_registrations = get_option( 'wpswmanager_registrations', array() );
+			$current_registrations = array_keys( $this->service_workers );
+			$areEqual = array_count_values( $last_registrations ) === array_count_values( $current_registrations );
+			if ( ! $areEqual ) {
+				$this->dynamic_server->invalidate_files( array( self::SW_REGISTRAR_SCRIPT_URL ) );
+				update_option( 'wpswmanager_registrations', $current_registrations );
+			}
+		}
+
+		public function enqueue_registrar() {
+			$url = WP_Serve_File::get_relative_to_wp_root_url( self::SW_REGISTRAR_SCRIPT_URL );
+			wp_enqueue_script( self::SW_REGISTRAR_SCRIPT, $url );
+		}
+
+		private function add_new_sw( $scope ) {
+			$virtual_url = "wpswmanager/sw/sw@$scope";
+			$real_url = $this->router->add_route( $virtual_url, array( $this, 'write_sw' ), $scope );
+			$service_worker = new WP_SW_Manager_Combinator( $real_url );
+			$this->service_workers[ $scope ] = $service_worker;
+		}
+
+		public function sw_registrar() {
+			$contents = file_get_contents( __DIR__ . '/lib/js/sw-registrar.js' );
+			$contents = str_replace( '$enabledSw', $this->json_for_sw_registrations(), $contents );
+			return array(
+				'content' => $contents,
+				'contentType' => 'application/javascript',
+			);
+		}
+
+		public function write_sw( $scope ) {
+			$service_worker = $this->service_workers[ $scope ];
+			header( 'Content-Type: application/javascript' );
+			header( 'Service-Worker-Allowed: ' . $scope );
+			include( __DIR__ . '/lib/js/localforage.nopromises.min.js' );
+			include( __DIR__ . '/lib/js/sw-base.js' );
+			include( __DIR__ . '/lib/js/sw-start.js' );
+			$service_worker->write_content();
+			$this->end();
+		}
+
+		private function end() {
+			wp_die();
+		}
+
+		private function json_for_sw_registrations() {
+			$registrations = array();
+			foreach ( $this->service_workers as $scope => $service_worker ) {
+				if ( $service_worker->has_content() ) {
+					$registrations[] = array(
+						'scope' => $scope,
+						'url' => $service_worker->get_url(),
+					);
+				}
+			}
+			return wp_json_encode( $registrations );
```

Comment:
Since we are supporting from WordPress 3.8, we can not use `wp_json_encode()`, at least, not for WP versions prior to 4.1.0.